src/hotspot/share/nmt/vmatree.hpp

```diff
@@ -0,0 +1,192 @@
+// Each of these points stores information about a state change.
+// For example, the state may go from released memory to committed memory,
+// or from committed memory of a certain MEMFLAGS to committed memory of a different MEMFLAGS.
+// The set of points is stored in a balanced binary tree for efficient querying and updating.
+class VMATree {
+  class AddressComparator {
+  public:
+    static int cmp(size_t a, size_t b) {
+      if (a < b) return -1;
+      if (a == b) return 0;
+      if (a > b) return 1;
+      ShouldNotReachHere();
+    }
+  };
+
+public:
+  enum class StateType : uint8_t { Reserved, Committed, Released };
+
+  // Each point has some stack and a flag associated with it.
+  struct Metadata {
+    NativeCallStackStorage::StackIndex stack_idx;
+    MEMFLAGS flag;
+
+    Metadata()
+      : stack_idx(),
+        flag(mtNone) {
+    }
+    Metadata(NativeCallStackStorage::StackIndex stack_idx, MEMFLAGS flag)
+      : stack_idx(stack_idx),
+        flag(flag) {
+    }
+    static bool equals(const Metadata& a, const Metadata& b) {
+      return NativeCallStackStorage::StackIndex::equals(a.stack_idx, b.stack_idx) &&
+             a.flag == b.flag;
+    }
+  };
+
+  struct IntervalState {
+  private:
+    // Store the type and flag as two bytes
+    uint8_t type_flag[2];
+    NativeCallStackStorage::StackIndex sidx;
+
+  public:
+    IntervalState() : type_flag{0,0}, sidx() {}
+    IntervalState(StateType type, Metadata data) {
+      type_flag[0] = static_cast<uint8_t>(type);
+      type_flag[1] = static_cast<uint8_t>(data.flag);
+      sidx = data.stack_idx;
+    }
+
+    StateType type() const {
+      return static_cast<StateType>(type_flag[0]);
+    }
+
+    MEMFLAGS flag() const {
+      return static_cast<MEMFLAGS>(type_flag[1]);
+    }
+
+    Metadata metadata() const {
+      return Metadata{sidx, flag()};
+    }
+  };
+
+  // An IntervalChange indicates a change in state between two intervals. The incoming state
+  // is denoted by in, and the outgoing state is denoted by out.
+  struct IntervalChange {
+    IntervalState in;
+    IntervalState out;
+
+    bool is_noop() {
+      return (in.type() == StateType::Released && out.type() == StateType::Released) ||
+             (in.type() == out.type() && Metadata::equals(in.metadata(), out.metadata()));
+    }
+  };
+
+  using VTreap = TreapNode<size_t, IntervalChange, AddressComparator>;
+  TreapCHeap<size_t, IntervalChange, AddressComparator> tree;
+
+  VMATree()
+    : tree() {
+  }
+
+  struct SingleDiff {
+    int64_t reserve;
+    int64_t commit;
+  };
+  struct SummaryDiff {
+    SingleDiff flag[mt_number_of_types];
+    SummaryDiff() {
+      for (int i = 0; i < mt_number_of_types; i++) {
+        flag[i] = SingleDiff{0, 0};
+      }
+    }
+  };
+
+  SummaryDiff register_mapping(size_t A, size_t B, StateType state, Metadata& metadata);
+
+  SummaryDiff reserve_mapping(size_t from, size_t sz, Metadata& metadata) {
```

Comment:
Let's wait with this until we actually port over the `VirtualMemoryTracker` to use `VMATree`.

Reply:
I think we should rethink recording specific stacks for uncommitted memory. I don't believe anyone cares who reserves uncommitted memory; or who uncommits memory. And this only leads to splintering the tree, if we uncommit from different callsites. We should consider keeping stacks for committed memory only, and use some noop stack placeholder for uncommitted mmeory.

Reply:
The issue, as I see it, is that we think of committing memory as a "layering" on top of reserving memory, and when that commit goes away the underlying layer of reserved memory is exposed again. In our VMATree, we don't store that underlying reservation anymore.

So what to do? If we add callstack and MEMFLAGS for uncommitting memory then that's an easy solution. The best would be to keep VMT's semantics here. We can do that, if the metadata stored is doubled in size per node and we recognise this pattern.

Still, I'll re-iterate: This is a problem for tomorrow, when we do port VMT.

Reply:
Note that the callstack information will be wrong for uncommitted memory though, since what should happen is that the now-again-reserved region should be re-accounted to the callstack that originally did the total reservation, not the callstack now doing the uncommitting.

But tbh, as much as I used NMT with customer cases over the years, the detail view with callstacks was never of much use for the mmap case anyway. It is very useful for mallocs, but for mmaps, I think it is already quite broken today, with reservations splintering up into many different accountings. I would not have a problem with just accounting stacks for reservations, not for commits.